tower-balance/src/p2c/service.rs

```diff
@@ -230,45 +287,58 @@ where
         // previously-selected `ready_index` if appropriate.
         self.poll_discover()?;
 
-        if let Some(index) = self.ready_index {
-            debug_assert!(index < self.endpoints.len());
-            // Ensure the selected endpoint is still ready.
-            match self.poll_endpoint_index_load(index) {
-                Ok(Async::Ready(_)) => return Ok(Async::Ready(())),
-                Ok(Async::NotReady) => {}
-                Err(e) => {
-                    drop(self.endpoints.swap_remove_index(index));
-                    info!("evicting failed endpoint: {}", e.into());
-                }
-            }
-
-            self.ready_index = None;
-        }
+        self.poll_unready();
```

Comment:
I wonder whether it'd be better to only `poll_unready` if there aren't enough things in `self.ready`?

Reply:
This adds some complexities that I'd prefer to defer for now. Specifically, this means that new instances added to the load balancer won't be considered until all existing nodes are exhausted. I think this would be surprising in practice and difficult for users to debug.

Instead, we'll probably want to consider an alternate _aperture_ load balancer at some point in the future, which works more as you describe -- fitting the number of nodes being considered to the load requirements.

As for cost of polling unready, in practice I think the endpoint stack should use a spawn-ready layer so that the balancer isn't responsible for driving all endpoints to readiness, and is only notified as they are fully ready.

Reply:
Yeah, that's a good point. In fact, it's similar to exactly what I ended up doing in [`tokio-tower`](https://github.com/tower-rs/tokio-tower) -- the service is spawned and provides a single-element queue of requests precisely so that the service is driven elsewhere.